### PR TITLE
[FLINK-12213][network] Pass TaskManagerMetricGroup into constructor of NetworkEnvironment

### PR DESCRIPTION
**Pull request by zhijiangW**
## What is the purpose of the change

*At the moment `NetworkEnvironment#getNetworkBufferPool` is called to add network related `MetricGroup`. In order to simplify the public API in `NetworkEnvironment` which is regarded as default `ShuffleService` implementation, we could pass the `TaskManagerMetricGroup` into constructor of `NetworkEnvironment`, then the related network `MetricGroup` could be added internally.*

## Brief change log

  - *Pass `TaskManagerMetricGroup` into constructor of `NetworkEnvironment`*
  - *Make `TaskManagerMetricGroup` as a component of `TaskManagerServices`*
  - *Fix the process of creating `TaskManagerMetricGroup` and related parts*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zhijiangW** [comment]
@flinkbot attention @azagrebin

**Comment by zhijiangW** [comment]
I would double check whether the `TaskManagerMetricGroup` should be close in `TaskManagerServices`, and atm it is closed by `TaskExecutor` to keep the same behavior during refactor.
Then I would refactor the related tests in the following commit to avoid calling `NetworkEnvironment#getNetworkBufferPool`, then we could remove this getter completely.

**Comment by zhijiangW** [comment]
Considering my above comments, I added the `TaskManagerMetricGroup#close` during `TaskManagerServices#shutdown` to keep consistent with other components in `TaskManagerServices`. Regarding the usages of `NetworkEnvironment#getNetworkBufferPool` in current two tests, one can be fixed during creation of partition/gate inside `NetworkEnvironment` in other PR, and another can also be avoided by introducing `NetworkEnvironmentBuilder` to give `NetworkBufferPool` explicitly outside which would be adjusted in a separate PR if necessary. @azagrebin

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java [hunk 1]
```diff
@@ -63,11 +70,19 @@
...
-	public NetworkEnvironment(NetworkEnvironmentConfiguration config, TaskEventPublisher taskEventPublisher) {
+	public NetworkEnvironment(
+			NetworkEnvironmentConfiguration config,
+			TaskEventPublisher taskEventPublisher,
+			TaskManagerMetricGroup taskManagerMetricGroup) {
 		this.config = checkNotNull(config);
 
 		this.networkBufferPool = new NetworkBufferPool(config.numNetworkBuffers(), config.networkBufferSize());
 
+		checkNotNull(taskManagerMetricGroup);
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: nit: could we move into a separate static private method?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java [hunk 2]
```diff
@@ -63,11 +70,19 @@
 
 	private boolean isShutdown;
 
-	public NetworkEnvironment(NetworkEnvironmentConfiguration config, TaskEventPublisher taskEventPublisher) {
+	public NetworkEnvironment(
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: From dependency injection perspective, could we add a static factory method here which would create all dependency components, like networkBufferPool, connectionManager, doing metrics registration etc? Then it could just call the private constructor passing all configured dependencies. I know we do not use this constructor at the moment but it would help to keep code cleaner in long term. At the end `ShuffleMaster.createShuffleService(flinkConfig)` could merge parsing `NetworkEnvironmentConfiguration` and creation of dependency components before creating `NetworkEnvironment` as `ShuffleService`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 3]
```diff
@@ -281,20 +283,7 @@ public void testShouldShutDownTaskManagerServicesInPostStop() throws Exception {
...
-			rpc,
-			taskManagerConfiguration,
-			haServices,
-			taskManagerServices,
-			heartbeatServices,
-			UnregisteredMetricGroups.createUnregisteredTaskManagerMetricGroup(),
-			null,
-			dummyBlobCacheService,
-			testingFatalErrorHandler);
+		final TaskExecutor taskManager = createTaskExecutor(taskManagerServices);
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: NIce refactoring for TaskExecutor creation, could you move it into a separate commit?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerLocation.java [hunk 4]
```diff
@@ -82,30 +82,9 @@ public TaskManagerLocation(ResourceID resourceID, InetAddress inetAddress, int d
...
-			fqdnHostName = this.inetAddress.getCanonicalHostName();
-		}
-		catch (Throwable t) {
-			LOG.warn("Unable to determine the canonical hostname. Input split assignment (such as " +
-					"for HDFS files) may be non-local when the canonical hostname is missing.");
-			LOG.debug("getCanonicalHostName() Exception:", t);
-			fqdnHostName = this.inetAddress.getHostAddress();
-		}
-		this.fqdnHostName = fqdnHostName;
+		this.fqdnHostName = getFqdnHostName(inetAddress);
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/TaskManagerLocation.java]: Also, nice simplification, but probably for a separate commit?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java [hunk 5]
```diff
@@ -208,6 +219,8 @@ public void shutDown() throws FlinkException {
 
 		taskEventDispatcher.clearAll();
 
+		taskManagerMetricGroup.close();
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: if we decide to keep `taskManagerMetricGroup` in `TaskManagerServices` then `taskManagerMetricGroup.close();` is not needed in `TaskExecutor.stopTaskExecutorServices` and probably as a field there at all?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 6]
```diff
@@ -357,15 +357,16 @@ public static TaskExecutor startTaskManager(
 		TaskManagerServices taskManagerServices = TaskManagerServices.fromConfiguration(
 			taskManagerServicesConfiguration,
 			resourceID,
+			remoteAddress,
+			metricRegistry,
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Why do we need to move `taskManagerMetricGroup` from `TaskExecutor` to `TaskManagerServices`? It seems it would be enough to just create `taskManagerMetricGroup` before `taskManagerServices` and pass it to `TaskManagerServices.fromConfiguration` and have less changes.

Also, I am wondering whether it would make sense to pull out the creation of `NetworkEnviroment` completely from `TaskManagerServicesConfiguration.fromConfiguration`/`TaskManagerServices.fromConfiguration` and pass it as a parameter to `TaskManagerServices.fromConfiguration`. We could migrate completely `NetworkEnviroment` creation into `NetworkEnviroment.fromConfiguration(configuration)` to get closer to `ShuffleMaster.createShuffleService(configuration)`. Probably, as a separate commit.

**Comment by azagrebin** [comment]
@zhijiangW 
As for tests, as long as the `getNetworkBufferPool` is not part shuffle API and used only in internal tests of `NetworkEnvironment`, it is fine for the scope of shuffle refactoring. Further on, we can of course simplify it as well or use the constructor with created dependencies as I suggested in one of my comments. Maybe, it could a separate commit here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 6 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Yes it also makes sense to create the `TaskManagerMetricGroup` before `TaskManagerServices`. I just thought it should be regarded as a component of `TaskManagerServices` like `TaskManagerLocation`. Nevertheless from the aspect of minimal change, I am willing to make it separate as you suggested. 

That is a good idea to extract the creation of `NetworkEnvironment` completely. I would do that in separate commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java (hunk 5 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java]: That is true. We decide not to make `TaskManagerMetricGroup` as one component of `TaskManagerServices`, so keep the previous behavior in `TaskExecutor`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 2 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: Yes, I also considered this idea before, but not sure when to bring this factory. I could try to do it in this PR.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 1 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: Sorry, what is moved into private method?

**Comment by zhijiangW** [comment]
@azagrebin  thanks for the efficient reviews! 
I agree with the point of internal testing which could retain the getter method in `NetworkEnvironment`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 1 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: private static registerNetworkMetrics(ComponentMetricGroup parentMetricGroup) {
  checkNotNull(parentMetricGroup);
  MetricGroup networkGroup = parentMetricGroup.addGroup(METRIC_GROUP_NETWORK);
  networkGroup.<Integer, Gauge<Integer>>gauge(METRIC_TOTAL_MEMORY_SEGMENT, 
  networkBufferPool::getTotalNumberOfMemorySegments);
  networkGroup.<Integer, Gauge<Integer>>gauge(METRIC_AVAILABLE_MEMORY_SEGMENT, 
  networkBufferPool::getNumberOfAvailableMemorySegments);
}

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java [hunk 7]
```diff
@@ -63,11 +70,19 @@
...
+			NetworkEnvironmentConfiguration config,
+			TaskEventPublisher taskEventPublisher,
+			TaskManagerMetricGroup taskManagerMetricGroup) {
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: btw, do we need `TaskManagerMetricGroup`? 
Could it be `ComponentMetricGroup parentMetricGroup` to have less dependency?

Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: Sure, we actually use the methods of `ComponentMetricGroup`  here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 6 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Considering the issue of `NetworkEnviroment.fromConfiguration(configuration)`, we should decouple the `NetworkEnvironmentConfiguration` from `TaskManagerServicesConfiguration` and the current creation of `NetworkEnvironmentConfiguration` also relies on other parameters such as `maxJvmHeapMemory`, `remoteAddress` and `localCommunication`.

So it might be involved in more refactoring to create `NetworkEnvironment` from flink `Configuration` directly. Then I want to create a separate ticket for this issue. What do you think?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 2 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: I guess your mentioned factory is similar with the `Builder` function to create every internal components or pass them from outsides as in test usages.

I thought of this issue from the motivation of tests aspect, but there seems only one `TaskEventDispatcher` component, so I have not focused on it before. Now I add the new component `ComponentMetricGroup`,  it seems necessary. 

My suggestion is providing the simple build method atm for avoiding creating these two components in related tests. The final form of factory is related to `NetworkEnvironment#fromConfiguration(flink configuration)` and `ShuffleMaster.createShuffleService(flinkConfig)`, so it might be better to consider them together when working the PR for `ShuffleMaster.createShuffleService(flinkConfig)`. What do you think?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 6 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: agreed 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 2 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: agreed, it is actually more related to the below comment about `NetworkEnvironment` creation refactoring in `TaskManagerServices`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 8]
```diff
@@ -1369,18 +1256,8 @@ public void testMaximumRegistrationDurationAfterConnectionLoss() throws Exceptio
...
-			rpc,
-			TaskManagerConfiguration.fromConfiguration(configuration),
-			haServices,
-			taskManagerServices,
-			new HeartbeatServices(heartbeatInterval, 10L),
-			UnregisteredMetricGroups.createUnregisteredTaskManagerMetricGroup(),
-			null,
-			dummyBlobCacheService,
-			testingFatalErrorHandler);
+		final TaskExecutor taskExecutor = createTaskExecutor(taskManagerServices);
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: here we had `new HeartbeatServices(10L, 10L)`, not `HEARTBEAT_SERVICES `, should we keep it?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java [hunk 9]
```diff
@@ -105,21 +96,11 @@ public void testCreationFromConfigDefault() throws Exception {
...
-			tmResourceID,
-			Executors.directExecutor(),
-			MEM_SIZE_PARAM,
-			MEM_SIZE_PARAM);
+		TaskManagerServices taskManagerServices = createTaskManagerServices(config);
 
 		TaskExecutorLocalStateStoresManager taskStateManager = taskManagerServices.getTaskManagerStateStore();
 
-		String[] tmpDirPaths = taskManagerServicesConfiguration.getTmpDirPaths();
+		String[] tmpDirPaths = ConfigurationUtils.parseTempDirectories(config);
```
Comment (azagrebin) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java]: Why is this change needed?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java [hunk 10]
```diff
@@ -79,9 +90,21 @@ public NetworkEnvironment(NetworkEnvironmentConfiguration config, TaskEventPubli
 
 		this.taskEventPublisher = checkNotNull(taskEventPublisher);
 
+		registerNetworkMetrics(parentMetricGroup);
+
 		isShutdown = false;
 	}
 
+	private void registerNetworkMetrics(ComponentMetricGroup parentMetricGroup) {
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: As it is used in constructor, could we convert it into a static method:
`private static void registerNetworkMetrics(ComponentMetricGroup parentMetricGroup, networkBufferPool)`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java [hunk 11]
```diff
@@ -287,4 +311,8 @@ public boolean isShutdown() {
 			return isShutdown;
 		}
 	}
+
+	public static NetworkEnvironment build(NetworkEnvironmentConfiguration configuration) {
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: If it is used only for tests, I would suggest to move to test packages. It could be `NetworkEnvironmentConfigurationBuilder.buildNetworkEnvironment()` or maybe we could just rename  `NetworkEnvironmentConfigurationBuilder` to `NetworkEnvironmentBuilder` if we do not need builder only for `NetworkEnvironmentConfiguration`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 12]
```diff
@@ -357,18 +359,21 @@ public static TaskExecutor startTaskManager(
...
 			rpcService.getExecutor(), // TODO replace this later with some dedicated executor for io.
 			EnvironmentInformation.getSizeOfFreeHeapMemoryWithDefrag(),
 			EnvironmentInformation.getMaxJvmHeapMemory());
 
-		TaskManagerMetricGroup taskManagerMetricGroup = MetricUtils.instantiateTaskManagerMetricGroup(
-			metricRegistry,
-			taskManagerServices.getTaskManagerLocation(),
-			taskManagerServices.getNetworkEnvironment(),
-			taskManagerServicesConfiguration.getSystemResourceMetricsProbingInterval());
+		Optional<Time> systemResourceProbeInterval = ConfigurationUtils.getSystemResourceMetricsProbingInterval(configuration);
```
Comment (azagrebin) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Should `systemResourceProbeInterval` be moved outside of `instantiateTaskManagerMetricGroup`? Could we avoid this change?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java (hunk 8 above)
Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: Yes, we should make it shorter time in this case, so pass custom heartbeat service here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java (hunk 9 above)
Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java]: In order to reuse the `TaskManagerServicesConfiguration` in `createTaskManagerServices`, otherwise  the creation of `TaskManagerServicesConfiguration` has to be outside of `createTaskManagerServices`. 

Comparing with reuse the creation of `TaskManagerServicesConfiguration`, I think it is cheaper to use `ConfigurationUtils.parseTempDirectories(config)` instead here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 10 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: ok

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 12 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: It is just for keeping the same behavior as before. In previous `instantiateTaskManagerMetricGroup`, it first instantiates network metrics and them instantiates system metrics, so I keep the same sequence.

I would double check whether it is no problem to adjust the sequence. If so we could avoid this change.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 11 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: Yes, I think it is better to rename `NetworkEnvironmentConfigurationBuilder` to `NetworkEnvironmentBuilder` and migrate this build there.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java (hunk 9 above)
Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java]: True, I have a little concern that we recreate it with a fixed. If the way of creation in `TaskManagerServicesConfiguration.fromConfiguration`, it can diverge with the test. Maybe break down `createTaskManagerServices(config)` into `createTaskManagerServicesConfiguration(config)`, `createTaskManagerServices(taskManagerServicesConfiguration)` and  in this case?

Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/TaskExecutorLocalStateStoresManagerTest.java]: Separating two methods might seem too thin for each method which only covers one sentence. 
Nevertheless I could accept both ways and separate it as you suggest. :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java [hunk 13]
```diff
@@ -63,7 +71,10 @@
 
 	private boolean isShutdown;
 
-	public NetworkEnvironment(NetworkEnvironmentConfiguration config, TaskEventPublisher taskEventPublisher) {
+	public NetworkEnvironment(
+			NetworkEnvironmentConfiguration config,
+			TaskEventPublisher taskEventPublisher,
+			ComponentMetricGroup parentMetricGroup) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: looks like this could be a plain `MetricGroup`

**Comment by zhijiangW** [comment]
Thanks for the reviews @azagrebin @zentol 

I have submitted and arranged the commits for addressing above comments.

**Comment by azagrebin** [comment]
@flinkbot attention @zentol

**Comment by azagrebin** [comment]
@flinkbot approve all

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironmentBuilder.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.io.network.netty.NettyConfig;
+import org.apache.flink.runtime.metrics.groups.UnregisteredMetricGroups;
+import org.apache.flink.runtime.taskmanager.NetworkEnvironmentConfiguration;
+
+/**
+ * Builder for the {@link NetworkEnvironment}.
+ */
+public class NetworkEnvironmentBuilder {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironmentBuilder.java]: should be moved to test sources?

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/util/MetricUtils.java [hunk 15]
```diff
@@ -93,30 +91,26 @@ public static JobManagerMetricGroup instantiateJobManagerMetricGroup(
...
-		instantiateNetworkMetrics(networkGroup, network);
-
 		if (systemResourceProbeInterval.isPresent()) {
 			instantiateSystemMetrics(taskManagerMetricGroup, systemResourceProbeInterval.get());
 		}
 		return taskManagerMetricGroup;
 	}
 
-	public static void instantiateStatusMetrics(
+	private static void instantiateStatusMetrics(
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/util/MetricUtils.java]: unrelated

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java [hunk 16]
```diff
@@ -1472,22 +1349,12 @@ public void testIgnoringSlotRequestsIfNotRegistered() throws Exception {
...
-			new TaskManagerServicesBuilder()
-				.setTaskSlotTable(taskSlotTable)
-				.setTaskManagerLocation(taskManagerLocation)
-				.build(),
-			new HeartbeatServices(heartbeatInterval, 1000L),
-			UnregisteredMetricGroups.createUnregisteredTaskManagerMetricGroup(),
-			null,
-			dummyBlobCacheService,
-			testingFatalErrorHandler);
+		final TaskExecutor taskExecutor = createTaskExecutor(new TaskManagerServicesBuilder()
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: Why are we no longer passing `new HeartbeatServices(heartbeatInterval, 1000L)`?

Comment (azagrebin) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: do you think we need always a new one? cannot we share `HEARTBEAT_SERVICES = new HeartbeatServices(1000L, 1000L);` as it is now in the PR?

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/util/MetricUtils.java (hunk 15 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/util/MetricUtils.java]: ok. I could revert it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java (hunk 16 above)
Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: I thought `HeartbeatServices`  could be reused in the tests only if it needs different internal and timeout in special tests. In this test the created new `HeartbeatServices` is the same with `HEARTBEAT_SERVICES`, so I reuse the static one. Referring some other previous tests, they also reuse the same service.  Do you have other concerns? @zentol

**Comment by zhijiangW** [comment]
I submitted a separate fixup commit for addressing comments except for the heartbeat issue which is under confirmation. @zentol

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java (hunk 16 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: I don't know whether we need a separate heartbeat service. My general approach to tests is to not change such details, as there could be repercussions one isn't aware of.

I'm not really concerned about correctness in this case; my point essentially boils down to "Had you not changed this detail we wouldn't even have to worry about these changes potentially affecting the test."

Comment (zhijiangW) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java]: Thanks for the explanation. 

Ok, that makes sense to keep the same behavior during refactoring. Even though it is not necessary to create a new heartbeat service as we thought, it should be in a separate ticket for this issue, not involved in my changes.
